compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/validator/ServiceValidator.java

```diff
@@ -18,98 +18,387 @@
+            Map<String, MethodSymbol> methods = classSymbol.methods();
+            for (MethodSymbol methodSymbol : methods.values()) {
+                Location methodLocation = getLocation(methodSymbol, location);
+                if (methodSymbol.kind() == SymbolKind.RESOURCE_METHOD) {
+                    resourceMethodFound = true;
+                    validateResourceMethod((ResourceMethodSymbol) methodSymbol, methodLocation);
+                } else if (isRemoteFunction(methodSymbol)) {
+                    updateContext(this.context, CompilationError.INVALID_FUNCTION, methodLocation);
                 }
             }
+            if (!resourceMethodFound) {
+                updateContext(this.context, CompilationError.MISSING_RESOURCE_FUNCTIONS, location);
+            }
+        }
+    }
+
+    private void validateReturnTypeUnion(UnionTypeSymbol unionTypeSymbol, Location location) {
+        List<TypeSymbol> effectiveTypes = getEffectiveTypes(unionTypeSymbol);
+        if (effectiveTypes.isEmpty()) {
+            updateContext(this.context, CompilationError.INVALID_RETURN_TYPE_ERROR_OR_NIL, location);
+        } else if (effectiveTypes.size() == 1) {
+            validateReturnType(effectiveTypes.get(0), location);
+        } else {
+            for (TypeSymbol typeSymbol : effectiveTypes) {
+                validateUnionTypeMember(typeSymbol, location);
+            }
+        }
+    }
+
+    private void validateUnionTypeMember(TypeSymbol memberType, Location location) {
+        if (!isDistinctServiceReference(memberType)) {
+            updateContext(this.context, CompilationError.INVALID_UNION_MEMBER_TYPE, location);
+        } else {
+            validateReturnType(memberType, location);
+        }
+    }
+
+    private void validateRecordFields(RecordTypeSymbol recordTypeSymbol, Location location) {
+        Map<String, RecordFieldSymbol> recordFieldSymbolMap = recordTypeSymbol.fieldDescriptors();
+        for (RecordFieldSymbol recordField : recordFieldSymbolMap.values()) {
+            validateReturnType(recordField.typeDescriptor(), location);
+            if (isInvalidFieldName(recordField.getName().orElse(""))) {
+                updateContext(this.context, CompilationError.INVALID_FIELD_NAME, location);
+            }
+        }
+    }
+
+    private boolean isContextParameter(TypeSymbol typeSymbol) {
+        if (typeSymbol.getName().isEmpty()) {
+            return false;
+        }
+        if (!isGraphqlModuleSymbol(typeSymbol)) {
+            return false;
+        }
+        return CONTEXT_IDENTIFIER.equals(typeSymbol.getName().get());
+    }
+
+    private boolean isFileUploadParameter(TypeSymbol typeSymbol) {
+        if (typeSymbol.getName().isEmpty()) {
+            return false;
+        }
+        if (!isGraphqlModuleSymbol(typeSymbol)) {
+            return false;
         }
-        return false;
+        return FILE_UPLOAD_IDENTIFIER.equals(typeSymbol.getName().get());
     }
 }
```

Comment:
if we have a distinct service class and a record type as union-type members, do we allow them?

Reply:
No, we don't. Only distinct service classes are allowed.